explorer/interpreter/type_checker.cpp

```diff
@@ -5186,9 +5186,24 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
     *trace_stream_ << "** declaring class " << class_decl->name() << "\n";
   }
   Nonnull<SelfDeclaration*> self = class_decl->self();
-
   ImplScope class_scope(scope_info.innermost_scope);
 
+  // The base class and member declarations may refer to the class, so we must
+  // set the static type before we start processing them. We can't set the
+  // constant value until later, but the base class declaration doesn't need it.
```

Comment:
It looks like the error we get in the new test is appearing because we try to look at the constant value when evaluating the base class declaration, and it's not there. Maybe that's worth mentioning here -- this comment makes it sound like it'd be fine if the constant value were set early, but I think our diagnosis of inheritance cycles currently relies on that not being done. And conversely, I think there are correct cases (eg, CRTP) when we would want to use the constant value of the derived class in the base class declaration that won't work because it's not set yet.

Reply:
Good point; I've added a check that the base class is complete, which should make this less brittle. I don't know if that handles the CRTP case, but we can cross that bridge when we come to it.